### PR TITLE
Add Admin API to Fetch Messages Within a Particular Window

### PR DESCRIPTION
**Pull request by dav-is**
This adds two new admin APIs that allow us to fetch messages from a room within a particular time.

It copies the logic from the [`/rooms/<roomId>/messages`](https://github.com/matrix-org/synapse/blob/390b7ce946173b41a61f427ef25a9a1d0371ad0b/synapse/rest/client/room.py#L623) and [`/rooms/<roomId>/timestamp_to_event`](https://github.com/matrix-org/synapse/blob/390b7ce946173b41a61f427ef25a9a1d0371ad0b/synapse/rest/client/room.py#L1233) endpoints into the admin API. When using the admin API, synapse should ignore the user's permissions other than the fact that they are an admin. This means the messages can be used for auditing purposes because no messages have been hidden.

This PR also adds [docs for these new APIs](https://github.com/dav-is/synapse/blob/dav-is/add-messages-admin-api/docs/admin_api/rooms.md#room-messages-api).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/rest/admin/rooms.py [hunk 1]
```diff
@@ -858,3 +859,108 @@ async def on_PUT(
...
+        msgs = await self.pagination_handler.get_messages(
+            room_id=room_id,
+            requester=requester,
+            pagin_config=pagination_config,
+            as_client_event=as_client_event,
+            event_filter=event_filter,
+            use_admin_priviledge=True,
+        )
+
+        return 200, msgs
```
Comment (dklimpel) [inline synapse/rest/admin/rooms.py:915]:
```suggestion
        return HTTPStatus.OK, msgs
```

File: tests/rest/admin/test_room.py [hunk 2]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+        user_tok = self.login("foo1", "pass")
+        room_id = self.helper.create_room_as(user, tok=user_tok)
+
+        token = "t1-0_0_0_0_0_0_0_0_0"
+        channel = self.make_request(
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/messages?access_token=%s&from=%s"
+            % (room_id, self.admin_user_tok, token),
+        )
+        self.assertEqual(HTTPStatus.OK, channel.code)
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1809]:
```suggestion
        self.assertEqual(200, channel.code)
```

There was a discussion and decision by Synapse Team to use in tests integers.

File: tests/rest/admin/test_room.py [hunk 3]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+        user_tok = self.login("foo2", "pass")
+        room_id = self.helper.create_room_as(user, tok=user_tok)
+
+        token = "s0_0_0_0_0_0_0_0_0"
+        channel = self.make_request(
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/messages?access_token=%s&from=%s"
+            % (room_id, self.admin_user_tok, token),
+        )
+        self.assertEqual(HTTPStatus.OK, channel.code)
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1826]:
```suggestion
        self.assertEqual(200, channel.code)
```

File: tests/rest/admin/test_room.py [hunk 4]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+    def test_stream_token_is_accepted_for_fwd_pagianation(self) -> None:
+        user = self.register_user("foo2", "pass")
+        user_tok = self.login("foo2", "pass")
+        room_id = self.helper.create_room_as(user, tok=user_tok)
+
+        token = "s0_0_0_0_0_0_0_0_0"
+        channel = self.make_request(
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/messages?access_token=%s&from=%s"
+            % (room_id, self.admin_user_tok, token),
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1824]:
```suggestion
            "/_synapse/admin/v1/rooms/%s/messages?from=%s"
            % (room_id, self.admin_user_tok, token),
            access_token=self.admin_user_tok
```

IMO the access_token should not be part of the URL for security reason (e.g. exposing in log files).

File: tests/rest/admin/test_room.py [hunk 5]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+        second_token_str = self.get_success(second_token.to_string(store))
+
+        # Send a third event in the room to ensure we don't fall under any edge case
+        # due to our marker being the latest forward extremity in the room.
+        self.helper.send(room_id, body="message 3", tok=user_tok)
+
+        # Check that we get the first and second message when querying /messages.
+        channel = self.make_request(
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/messages?access_token=%s&from=%s&dir=b&filter=%s"
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1866]: The same with `access_token`.

File: tests/rest/admin/test_room.py [hunk 6]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+            % (
+                room_id,
+                self.admin_user_tok,
+                second_token_str,
+                json.dumps({"types": [EventTypes.Message]}),
+            ),
+        )
+        self.assertEqual(channel.code, HTTPStatus.OK, channel.json_body)
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1874]:
```suggestion
        self.assertEqual(channel.code, 200, channel.json_body)
```

File: tests/rest/admin/test_room.py [hunk 7]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1903]:
```suggestion
        self.assertEqual(channel.code, 200, channel.json_body)
```

File: tests/rest/admin/test_room.py [hunk 8]
```diff
@@ -1792,6 +1795,134 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/messages?access_token=%s&from=%s&dir=b&filter=%s"
+            % (
+                room_id,
+                self.admin_user_tok,
+                first_token_str,
+                json.dumps({"types": [EventTypes.Message]}),
+            ),
+        )
+        self.assertEqual(channel.code, HTTPStatus.OK, channel.json_body)
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1921]:
```suggestion
        self.assertEqual(channel.code, 200, channel.json_body)
```

File: synapse/rest/admin/rooms.py [hunk 9]
```diff
@@ -858,3 +859,108 @@ async def on_PUT(
...
+
+    PATTERNS = admin_patterns("/rooms/(?P<room_id>[^/]*)/messages$")
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
+        self._hs = hs
+        self.clock = hs.get_clock()
+        self.pagination_handler = hs.get_pagination_handler()
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastores().main
```
Comment (dklimpel) [inline synapse/rest/admin/rooms.py:877]:
```suggestion
        self._clock = hs.get_clock()
        self._pagination_handler = hs.get_pagination_handler()
        self._auth = hs.get_auth()
        self._store = hs.get_datastores().main
```

File: synapse/rest/admin/rooms.py [hunk 10]
```diff
@@ -858,3 +859,108 @@ async def on_PUT(
...
+
+class RoomMessagesRestServlet(RestServlet):
+    """
+    Get messages list of a room.
+    """
+
+    PATTERNS = admin_patterns("/rooms/(?P<room_id>[^/]*)/messages$")
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
```
Comment (dklimpel) [inline synapse/rest/admin/rooms.py:872]:
```suggestion

```
Is it needed?

File: synapse/rest/admin/rooms.py [hunk 11]
```diff
@@ -858,3 +859,107 @@ async def on_PUT(
...
+    GET /_synapse/admin/v1/rooms/<roomID>/timestamp_to_event?ts=<timestamp>&dir=<direction>
+    {
+        "event_id": ...
+    }
+    """
+
+    PATTERNS = admin_patterns("/rooms/(?P<room_id>[^/]*)/timestamp_to_event$")
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
```
Comment (dklimpel) [inline synapse/rest/admin/rooms.py:941]:
```suggestion

```

File: tests/rest/admin/test_room.py [hunk 12]
```diff
@@ -1792,6 +1795,159 @@ def test_get_joined_members_after_leave_room(self) -> None:
         self.assertEqual(403, channel.code, msg=channel.json_body)
         self.assertEqual(Codes.FORBIDDEN, channel.json_body["errcode"])
 
+    def test_timestamp_to_event(self) -> None:
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1798]: It probably makes sense to create an own `class` for these tests. `RoomMessageAndTimestampTestCase` is an example. Perhaps there is a better name for this.

```suggestion

class RoomMessageAndTimestampTestCase(unittest.HomeserverTestCase):
    servlets = [
        synapse.rest.admin.register_servlets,
        room.register_servlets,
        login.register_servlets,
    ]

    def prepare(self, reactor: MemoryReactor, clock: Clock, hs: HomeServer) -> None:
        self.user = self.register_user("foo", "pass")
        self.user_tok = self.login("foo", "pass")
        self.room_id = self.helper.create_room_as(user, tok=user_tok)

    def test_timestamp_to_event(self) -> None:
```

Than you do not need to recreate the user and room in every test case.

File: tests/rest/admin/test_room.py [hunk 13]
```diff
@@ -1792,6 +1795,159 @@ def test_get_joined_members_after_leave_room(self) -> None:
...
+        ts = str(round(time.time() * 1000))
+
+        channel = self.make_request(
+            "GET",
+            "/_synapse/admin/v1/rooms/%s/timestamp_to_event?dir=b&ts=%s"
+            % (room_id, ts),
+            access_token=self.admin_user_tok,
+        )
+        self.assertEqual(200, channel.code)
+        self.assertTrue("event_id" in channel.json_body)
```
Comment (dklimpel) [inline tests/rest/admin/test_room.py:1817]:
```suggestion
        self.assertIn("event_id", channel.json_body)
```

Is it the same?

File: docs/admin_api/rooms.md [hunk 14]
```diff
@@ -393,6 +393,151 @@ A response body like the following is returned:
...
+The following path parameters are required:
+
+* `room_id` - The ID of the room you wish to check.
+
+The following query parameters are available:
+
+* `ts` - a timestamp in milliseconds where we will find the closest event in
+  the given direction.
+* `dir` - can be `f` or `b` to indicate forwards and backwards in time from the
+  given timestamp.
```
Comment (dklimpel) [inline docs/admin_api/rooms.md:535]:
```suggestion
  given timestamp. Defaults to `f`.
```

File: synapse/handlers/pagination.py [hunk 15]
```diff
@@ -561,9 +572,12 @@ async def get_messages(
...
 
-        aggregations = await self._relations_handler.get_bundled_aggregations(
-            events, user_id
-        )
+        if not use_admin_priviledge:
+            aggregations = await self._relations_handler.get_bundled_aggregations(
+                events, user_id
+            )
+        else:
+            aggregations = None  # TODO: an admin might want aggregations
```
Comment (erikjohnston) [inline synapse/handlers/pagination.py:580]: Yeah, I think we should always bundle the aggregations. The `user_id` is merely used to filter out ignored users, so it's safe to just pass in the admin user ID.

**Review by erikjohnston — COMMENTED** [review]
Otherwise looks good to me!

**Review by erikjohnston — APPROVED** [review]
Woot! Thanks

**Pull request merged**